internal/pkg/describe/app.go

```diff
@@ -90,5 +100,39 @@ func NewAppDescriber(opt NewAppDescriberConfig) (*AppDescriber, error) {
 //
 // If the Version field does not exist, then it's a legacy template and it returns an deploy.LegacyAppTemplateVersion and nil error.
 func (d *AppDescriber) Version() (string, error) {
-	return deploy.LegacyAppTemplateVersion, nil
+	metadata := struct {
+		TemplateVersion string `yaml:"TemplateVersion"`
+	}{}
+
+	appStackName := stack.NameForAppStack(d.app)
+	appStackMetadata, err := d.cfn.Metadata(cloudformation.WithStackName(appStackName))
+	if err != nil {
+		return "", fmt.Errorf("get metadata for app stack %s: %w", appStackName, err)
+	}
+	if err := yaml.Unmarshal([]byte(appStackMetadata), &metadata); err != nil {
+		return "", fmt.Errorf("unmarshal Metadata property for app stack %s: %w", appStackName, err)
+	}
+	appStackVersion := metadata.TemplateVersion
+	if appStackVersion == "" {
+		appStackVersion = deploy.LegacyAppTemplateVersion
+	}
+
+	appStackSetName := stack.NameForAppStackSet(d.app)
+	appStackSetMetadata, err := d.cfn.Metadata(cloudformation.WithStackSetName(appStackSetName))
+	if err != nil {
+		return "", fmt.Errorf("get metadata for app stack set %s: %w", appStackSetName, err)
+	}
+	if err := yaml.Unmarshal([]byte(appStackSetMetadata), &metadata); err != nil {
+		return "", fmt.Errorf("unmarshal Metadata property for app stack set %s: %w", appStackSetName, err)
+	}
+	appStackSetVersion := metadata.TemplateVersion
+	if appStackSetVersion == "" {
```

Comment:
This can only evaluate to true if the app version is a legacy template and the stackset template is legacy, is that intended? (l.112 already modifies the `&metadata.TemplateVersion` field)

Reply:
Hmm i don't think i fully understand but yeah this is just for legacy template so then we set appStackSetVersion to be the legacy version, which means the stackset is using legacy template.

Reply:
Let me try to rephrase it 😄  If the `appStackMetadata` is not `""` but the `appStackSetMetadata` is empty then this won't evaluate to true. Does that make sense ? :P

Reply:
yeah. what's the problem 🤔 then `appStackSetVersion` will be v0.0.0 and `appStackVersion` will be whatever. And it returns v0.0.0. It's different from env upgrade that for app upgrade updating from legacy does not need it's own method.

Reply:
Discussed offline. Very good callout: tested and found out `yaml.Unmarshal` won't overwrite the `metadata` object so we can't reuse the same object for both stack and stackset.